sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,196 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_PADCTRL_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_PADCTRL_H_
+
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - Should be autogenerated.
+// - Should padctrl/pinmux share some defines/constants (such as MIO_NUM) or do
+// we just autogen the same number twice?
+#define DIF_PADCTRL_DIO_NUM 32
+#define DIF_PADCTRL_MIO_NUM 32
+#define DIF_PADCTRL_ATTR_WIDTH 8
+#define DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH 2
+
+/**
+ * Pad attributes
+ *
+ * Some attributes have defined meanings, others are implementation defined
+ * (|kDifPadCtrlAttrAdditionalFirst| and onwards). not all pads will support all
+ * attributes.
+ */
+typedef enum dif_padctrl_attr {
+  kDifPadctrlAttrIOInvert = 0,
+  kDifPadctrlAttrOpenDrain,
+  kDifPadctrlAttrPullDown,
+  kDifPadctrlAttrPullUp,
+  kDifPadctrlAttrKeeper,
+  kDifPadctrlAttrWeakDrive
+  kDifPadctrlAttrAdditionalFirst = 0,
+  kDifPadctrlAttrAdditionalLast =
+    DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH == 0 ?
+      kDifPadctrlAdditionalAttrFirst :
+      kDifPadctrlAdditionalAttrFirst + DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH - 1,
+} dif_padctrl_attr_t;
+
+/**
+ * DIO pads
+ */
+typedef enum dif_padctrl_dio {
+  kDifPadctrlDioFirst = 0,
+  kDifPadctrlDioLast = DIF_PADCTRL_DIO_NUM - 1,
+} dif_padctrl_dio_t;
+
+/**
+ * MIO Pads
+ */
+typedef enum dif_padctrl_mio {
+  kDifPadctrlMioFirst = 0,
+  kDifPadctrlMioLast = DIF_PADCTRL_MIO_NUM - 1,
+} dif_padctrl_mio_t;
```

Comment:
There's two options for using these. Either:
- We end up casting from ints to these types, and use the First/Last to validate the range; or
- We expand the cases to support the number of Mios/Dios that we're expecting, ie `kDifPadctrlMio0`..`kDifPadctrlMio16`.

The latter is better for type-safety, but potentially not very configurable (this is a wider issue with parameterisation that I intend to address soon). I'm happy for the intention to be to stick to the former at the moment.

Reply:
I'll leave it this way for now